### DISCUSSION TITLE
FP sum of values in attributes: value generated with formulas and manually, use of icons

### DISCUSSION THREAD

**Discussion by Illegit** [freeplane/freeplane#1782] [Help] [answered]
Hi, I have a map where I need to sum the attribute values called "upd". It is a % of advancement.
![Screenshot 2024-04-14 100848](https://github.com/freeplane/freeplane/assets/62674157/e0890bf5-fe0f-41db-9ad5-1b65372ef5c8)
[Sum of UPD attribute.zip](https://github.com/freeplane/freeplane/files/14969964/Sum.of.UPD.attribute.zip)
I use icons: 1) the green tick to attribute te value 1 to upd and 2) the Zzz icon to tell to the formula not to account in upd that node.

Now the question I cannot answer: my formula works well always except when it is accounting a node with the same formula inside the upd value generating an error: which is the mistake? How to correct this formula?
![Screenshot 2024-04-14 101810](https://github.com/freeplane/freeplane/assets/62674157/910317f3-6a1c-4b2d-9689-4cf4c949a980)
you can play with the above map to test your solutions.
Thanks in any case for the attention.

**Comment by Illegit** [comment]
I tried a different formula, but I got the same problem as above: see the zip file
[sum of attr values w icons.zip](https://github.com/freeplane/freeplane/files/14971579/sum.of.attr.values.w.icons.zip)
the formula:
  `  if (children.size()<1){res=0}else{num=children.collect {if(!it.icons.contains("MarkdownHelper/completed")&!it.icons.contains("Notes/zzz")){it.attributes.findAll(){it.key=='upd'}.value}else if(it.icons.contains("MarkdownHelper/completed")&!it.icons.contains("Notes/zzz")){1}else{0}}.flatten().sum(0); den=node.children.count{!it.icons.contains("Notes/zzz")}; res=num/den}; format(res,'#.00')`
       
 this formula fails when summing up nodes with the same formula inside their upd attribute value
            what is wrong?
            how to correct it?
Thanks for feedback.

**Comment by macmarrum** [comment, marked as answer]
In the following code, `.value` gets you the raw value of the attribute, which in your case is the formula's text.
```groovy
            def updAttribute = child.attributes.find { it.key == 'upd' }
            return updAttribute ? updAttribute.value : 0
```
To get the result of the formula evaluation, you can use either `.getTransformed()` or `.findValues` 
```groovy
            def updAttribute = child.attributes.transformed.find { it.key == 'upd' }
            return updAttribute ? updAttribute.value : 0
```
`.findVlaues` will return a list of `Convertible`
```groovy
            def updAttributeList = child.attributes.findValues { key, value -> key == 'upd' }
            return updAttributeList ? updAttributeList[0].num : 0
```

The above can be simplified. `NodeRO` returns an attribute's value as `Convertible` when square brackets are used  => `node['upd']`.
Then you can use `.num0` on the `Convertible` to get it as a number; if not-a-number (e.g. null), it will return 0
```groovy
            return child['upd'].num0
```

**Comment by Illegit** [comment]
Thank you macmarrum, it works finely.
